python/tskit/tables.py

```diff
@@ -43,16 +43,55 @@
 import tskit.util as util
 from tskit import UNKNOWN_TIME
 
+
 dataclass_options = {"frozen": True}
 
 
+class LazyMetadata:
+    __slots__ = ["table", "_metadata", "_decoded"]
+
+
+class CachedMetadata:
+    """
+    Descriptor for lazy decoding of metadata on attribute access.
+    """
+
+    def __get__(self, row, owner):
+        if not row._decoded and row.table is not None:
+            builtins.object.__setattr__(
+                row, "_metadata", row.table.metadata_schema.decode_row(row._metadata)
+            )
+            builtins.object.__setattr__(row, "_decoded", True)
+        return row._metadata
+
+    def __set__(self, row, value):
+        builtins.object.__setattr__(row, "_decoded", False)
+        builtins.object.__setattr__(row, "_metadata", value)
+
+
+def lazy_metadata(cls):
+    """
+    Assign table reference to row dataclasses and apply caching descriptor
+    """
+    wrapped_init = cls.__init__
+
+    def new_init(self, *args, table=None, **kwargs):
+        builtins.object.__setattr__(self, "table", table)
+        wrapped_init(self, *args, **kwargs)
+
+    cls.__init__ = new_init
+    cls.metadata = CachedMetadata()
+    return cls
+
+
+@lazy_metadata
```

Comment:
Not obvious here why we need both a decorator and to inherit from LazyMetadata - one or other is easy to follow for a reader, but both is confusing. What's the rationale here?

Reply:
I guess we should spell it out up here where these classes are defined anyway, need to encode this knowledge somewhere.

Reply:
I couldn't figure out how to make an extra slot with a class decorator, annoying!

Reply:
Should we bother with the decorator then? We should be able to implement this in a straighforward inheritancy type way, right?

Reply:
Manged to do it all in a decorator in the end, can't do inheritance due to intercepting the constructor.